File: backend/models/novelModel.js
// Farid Dhiya Fairuz
// 247006111058
// B

const db = require('../config/db');

const NovelModel = {
  findAll: async () => {
    const [rows] = await db.execute('SELECT * FROM novels ORDER BY created_at DESC');
    return rows;
  },
  findById: async (id) => {
    const [rows] = await db.execute('SELECT * FROM novels WHERE id = ?', [id]);
    return rows[0] || null;
  },
  create: async ({ title, author, genre, chapters, rating }) => {
    const [result] = await db.execute(
      'INSERT INTO novels (title, author, genre, chapters, rating) VALUES (?, ?, ?, ?, ?)',
      [title, author, genre, chapters, rating]
    );
    return result.insertId;
  },
  update: async (id, { title, author, genre, chapters, rating }) => {
    const [result] = await db.execute(
      'UPDATE novels SET title = ?, author = ?, genre = ?, chapters = ?, rating = ? WHERE id = ?',
      [title, author, genre, chapters, rating, id]
    );
    return result.affectedRows;
  },
  delete: async (id) => {
    const [result] = await db.execute('DELETE FROM novels WHERE id = ?', [id]);
    return result.affectedRows;
  },
};

module.exports = NovelModel;

File: flutter_app/lib/models/novel_model.dart
// Farid Dhiya Fairuz - 247006111058 - B

class Novel {
  final int id;
  final String title;
  final String author;
  final String genre;
  final int chapters;
  final double rating;
  final String? createdAt;

  Novel({required this.id, required this.title, required this.author,
      required this.genre, required this.chapters, required this.rating, this.createdAt});

  factory Novel.fromJson(Map<String, dynamic> json) {
    return Novel(
      id: json['id'] is int ? json['id'] : int.tryParse(json['id'].toString()) ?? 0,
      title: json['title'] ?? '',
      author: json['author'] ?? '',
      genre: json['genre'] ?? '',
      chapters: json['chapters'] is int ? json['chapters'] : int.tryParse(json['chapters'].toString()) ?? 0,
      rating: json['rating'] is double ? json['rating'] : double.tryParse(json['rating'].toString()) ?? 0.0,
      createdAt: json['created_at'],
    );
  }
}
